functors task (2)/student/src/impl.cpp

```diff
@@ -5,50 +5,41 @@ void DataBrowser::userLeave(const std::string &userId)
     m_dataReaders.erase(userId);
 }
 
+void DataBrowser::userEnter(const std::string& userId)
+{
+    // Sorry, couldn't see it in header.
+    m_dataReaders.emplace(userId, getSelector());
+}
+
 bool DataBrowser::getDataType1(const std::string &userId, std::vector<size_t> &returnValues) const
 {
-    const auto& it = m_dataReaders.find(userId);
-    if (it == m_dataReaders.cend())
-    {
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType1(returnValues, 0);
+    auto getDataTypeAdapter = [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        using namespace std::placeholders;
+        return invokeDataRequest(
+            std::bind(&IDataSelector::getDataType1, _1, _2, 0), selector, returnValues);
+    };
+    return safeCall(userId, getDataTypeAdapter);
 }
 
 bool DataBrowser::getDataType2(std::vector<size_t> &returnValues, const std::string &userId) const
 {
-    const auto& it = m_dataReaders.find(userId);
-    if (it == m_dataReaders.cend())
-    {
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType2(returnValues);
+    auto getDataTypeAdapter = [this, &returnValues](const std::unique_ptr<IDataSelector>& selector) {
+        return invokeDataRequest(
+            std::mem_fn(&IDataSelector::getDataType2), selector, returnValues);
```

Comment:
_lambda is more explicit_

Reply:
Implemented three different variants to see the difference.

Reply:
Nice!

Reply:
The idea about lambdas is clear. Used std::bind() to see visual difference. Anything else to fix?

Reply:
it's ok now